instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/ReplayLocationTest.kt

```diff
@@ -0,0 +1,139 @@
+import com.mapbox.navigation.testing.ui.utils.getMapboxAccessTokenFromResources
+import com.mapbox.navigation.testing.ui.utils.runOnMainSync
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.flow.first
+import org.junit.Assert.assertEquals
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+
+@OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
+class ReplayLocationTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.java) {
+
+    private lateinit var mapboxNavigation: MapboxNavigation
+    private val tolerance = 0.000001
+
+    @get:Rule
+    val mapboxNavigationRule = MapboxNavigationRule()
+
+    private val realLocation = ApproximateCoordinates(0.9, 0.9, tolerance)
+
+    override fun setupMockLocation(): Location = mockLocationUpdatesRule.generateLocationUpdate {
+        latitude = realLocation.latitude
+        longitude = realLocation.longitude
+    }
+
+    @Before
+    fun setUp() {
+        runOnMainSync {
+            mapboxNavigation = MapboxNavigationProvider.create(
+                NavigationOptions.Builder(activity)
+                    .accessToken(getMapboxAccessTokenFromResources(activity))
+                    .build()
+            )
+        }
+    }
+
+    @Test
+    fun last_location_is_cleared_when_session_is_stopped() = sdkTest {
+        val firstReplayApproximateLocation = ApproximateCoordinates(1.0, 1.0, tolerance)
+        val secondReplayApproximateLocation = ApproximateCoordinates(1.2, 1.2, tolerance)
+        val rawLocations = mutableListOf<ApproximateCoordinates>()
+        val locationObserver = object : LocationObserver {
+            override fun onNewRawLocation(rawLocation: Location) {
+                rawLocations.add(
+                    ApproximateCoordinates(
+                        rawLocation.latitude,
+                        rawLocation.longitude,
+                        tolerance
+                    )
+                )
+            }
+
+            override fun onNewLocationMatcherResult(locationMatcherResult: LocationMatcherResult) {
+            }
+        }
+        mapboxNavigation.registerLocationObserver(locationObserver)
+        mapboxNavigation.startReplayTripSession()
+        updateReplayLocation(
+            mockLocationUpdatesRule.generateLocationUpdate {
+                latitude = firstReplayApproximateLocation.latitude
+                longitude = firstReplayApproximateLocation.longitude
+            }
+        )
+        mapboxNavigation.rawLocationUpdates()
+            .filter {
+                ApproximateCoordinates(
+                    it.latitude,
+                    it.longitude,
+                    tolerance
+                ) == firstReplayApproximateLocation
+            }
+            .first()
+        mapboxNavigation.stopTripSession()
+        assertEquals(List(rawLocations.size) { firstReplayApproximateLocation }, rawLocations)
+        rawLocations.clear()
+        delay(1000)
```

Comment:
Why do you need delay here? Is it going to stop working without it?
```suggestion
```

Reply:
Oooh, that's a nice catch! I think it was left here by accident, but when I removed it, the test started failing.
And it unveiled an interesting race condition. See the new commit for details.